### PR TITLE
executor: remove useless channel and fetch directly from copIterator

### PR DESCRIPTION
**Pull request by ichn-hu**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/11501, simplify the codebase with potential performance improvement.

In the original implementation, an additional channel is employed to buffer the data from copIterator, however it seems useless and adds more complexity to the codebase of select_result.

After removing the channel, there is also potential improvement in memory usage. If the amount of data is large, and the reading is slow, the channel will buffer data that causes extra memory overhead. In the following benchmark, a table of 1600 rows, each row of size 6M is created, using chunk size of 32 and DistSQLConcurrency = 15, if the mysql client reads slowly from the table, the original version cost almost twice the memory than the improved one.

Original:

image

Improved:
image

Sysbench also shows no performance regression.

### What is changed and how it works?

Refactor distsql/select_result.go, remove result channel and fetch data directly from copIterator.

### Check List 

Tests 

 - Integration test
 - No code

No visible changes.

No Side effects

No Related changes

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

**Comment by ichn-hu** [comment]
@SunRunAway PTAL

File: distsql/select_result.go [hunk 1]
```diff
@@ -85,74 +77,51 @@ type selectResult struct {
...
+	if resultSubset == nil {
+		r.selectResp = nil
+		return nil
+	}
+	r.selectResp = new(tipb.SelectResponse)
+	err = r.selectResp.Unmarshal(resultSubset.GetData())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	r.selectRespSize = r.selectResp.Size()
```
Comment (SunRunAway) [inline distsql/select_result.go]: We should keep tracking the memory of `r.selectResp()`.

Comment (ichn-hu) [reply distsql/select_result.go]: Updated, PTAL

File: distsql/select_result.go [hunk 2]
```diff
@@ -85,74 +77,55 @@ type selectResult struct {
...
 	if r.selectResp == nil || r.respChkIdx == len(r.selectResp.Chunks) {
-		err := r.getSelectResp()
-		if err != nil || r.selectResp == nil {
+		err := r.fetchResp(ctx)
+		if err != nil {
 			return err
 		}
+		if r.selectResp == nil {
+			return nil
+		}
```
Comment (wshwsh12) [inline distsql/select_result.go]: What's the difference between this code L123-128 and the old code L153-155?

File: distsql/select_result.go [hunk 3]
```diff
@@ -85,74 +77,55 @@ type selectResult struct {
...
+	}
+	r.updateCopRuntimeStats(resultSubset.GetExecDetails().CalleeAddress, resultSubset.RespTime())
+	r.feedback.Update(resultSubset.GetStartKey(), r.selectResp.OutputCounts)
 	r.partialCount++
-	r.feedback.Invalidate()
-	if re.result != nil && re.err == nil {
-		data = re.result.GetData()
+	sc.MergeExecDetails(resultSubset.GetExecDetails(), nil)
+	if len(r.selectResp.Chunks) == 0 {
+		return r.fetchResp(ctx)
```
[the remark this answers is not in the document]
Comment (ichn-hu) [reply distsql/select_result.go]: Sounds good, nice point.

**Comment by ichn-hu** [comment]
/bench

```
workload:
  threads: 512
```

**Comment by ichn-hu** [comment]
@wshwsh12 PTAL again

File: distsql/select_result.go [hunk 4]
```diff
@@ -85,74 +77,57 @@ type selectResult struct {
...
-			err := fmt.Errorf("%v", c)
-			logutil.Logger(ctx).Error("OOM", zap.Error(err))
-			r.results <- resultWithErr{err: err}
-		}
-
-		close(r.results)
-		duration := time.Since(startTime)
-		// TODO: Add a label to distinguish between success or failure.
-		// https://github.com/pingcap/tidb/issues/11397
-		metrics.DistSQLQueryHistgram.WithLabelValues(r.label, r.sqlType).Observe(duration.Seconds())
```
Comment (crazycs520) [inline distsql/select_result.go]: Please keep this metric alive.

Comment (SunRunAway) [reply distsql/select_result.go]: @ichn-hu Could you fix it?

**Pull request merged**